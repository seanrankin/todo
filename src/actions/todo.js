export const CREATE_TODO = 'CREATE_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const DESTROY_TODO = 'DESTROY_TODO';

export const createToDo = todo => ({
  type: CREATE_TODO,
  todo,
});

export const toggleToDo = (id, complete) => ({
  type: TOGGLE_TODO,
  id,
  complete,
});

export const destroyToDo = id => ({
  type: DESTROY_TODO,
  id,
});
