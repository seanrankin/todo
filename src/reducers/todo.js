import { CREATE_TODO, DESTROY_TODO, TOGGLE_TODO } from '../actions/todo';

export const initialState = {
  todos: [
    {
      id: 1,
      label: 'Start looking for a new job',
      complete: false,
      created: '2018-11-11T15:48:59.004Z',
    },
    {
      id: 2,
      label: 'Take the dog to the vet',
      complete: true,
      created: '2018-11-12T15:48:59.004Z',
    },
    {
      id: 3,
      label: 'Build Lego robot with Lily and Lucy',
      complete: false,
      created: '2018-11-13T15:48:59.004Z',
    },
    {
      id: 4,
      label: 'Plan the family holiday trip',
      complete: false,
      created: '2018-11-14T15:48:59.004Z',
    },
  ],
};

export default function todos(state = initialState, action) {
  switch (action.type) {
    case CREATE_TODO:
      return {
        ...state,
        todos: [...state.todos, action.todo],
      };
    case TOGGLE_TODO:
      return {
        ...state,
        todos: toggleToDo(state.todos, action.id, action.complete),
      };
    case DESTROY_TODO:
      return {
        ...state,
        todos: state.todos.filter(todo => todo.id !== action.id),
      };
    default:
      return state;
  }
}

export const toggleToDo = (todos, id, complete) => {
  return todos.map(item => {
    if (item.id !== id) {
      return item;
    }
    return {
      ...item,
      complete: !complete,
    };
  });
};
