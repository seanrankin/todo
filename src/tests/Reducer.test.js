import { CREATE_TODO, DESTROY_TODO, TOGGLE_TODO } from '../actions/todo';
import reducer, { initialState, toggleToDo } from '../reducers/todo';

import { todo } from './mocks/todo';

describe('ToDo reducer reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
    expect(initialState).toMatchSnapshot();
  });

  it('should handle CREATE_TODO', () => {
    const successAction = {
      type: CREATE_TODO,
      todo: todo,
    };

    expect(reducer(initialState, successAction)).toEqual({
      todos: [...initialState.todos, todo],
    });
  });

  it('should handle DESTROY_TODO', () => {
    const successAction = {
      type: DESTROY_TODO,
      id: 3,
    };

    expect(reducer(initialState, successAction)).toEqual({
      todos: [...initialState.todos.filter(todo => todo.id !== 3)],
    });
  });

  it('should handle TOGGLE_TODO', () => {
    const successAction = {
      type: TOGGLE_TODO,
      id: 5,
    };

    expect(reducer({ todos: [todo] }, successAction)).toEqual({
      todos: [
        {
          id: 5,
          label: 'Add this new todo',
          complete: true,
          created: '2018-11-17T15:48:59.004Z',
        },
      ],
    });
  });
});

describe('The toggleToDo reducer function', () => {
  it('should update the todos gracefully without mangling the order', () => {
    const toDoToMarkComplete = {
      id: 6,
      label: 'Add this new todo',
      complete: true,
      created: '2018-11-17T15:48:59.004Z',
    };
    expect(toggleToDo([todo, toDoToMarkComplete], 6, false)).toEqual([
      todo,
      {
        id: 6,
        label: 'Add this new todo',
        complete: true,
        created: '2018-11-17T15:48:59.004Z',
      },
    ]);
  });
});
