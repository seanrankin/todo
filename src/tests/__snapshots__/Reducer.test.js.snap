// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo reducer reducer should return the initial state 1`] = `
Object {
  "todos": Array [
    Object {
      "complete": false,
      "created": "2018-11-11T15:48:59.004Z",
      "id": 1,
      "label": "Start looking for a new job",
    },
    Object {
      "complete": true,
      "created": "2018-11-12T15:48:59.004Z",
      "id": 2,
      "label": "Take the dog to the vet",
    },
    Object {
      "complete": false,
      "created": "2018-11-13T15:48:59.004Z",
      "id": 3,
      "label": "Build Lego robot with Lily and Lucy",
    },
    Object {
      "complete": false,
      "created": "2018-11-14T15:48:59.004Z",
      "id": 4,
      "label": "Plan the family holiday trip",
    },
  ],
}
`;
